File: 0725 Split Linked List in Parts/0725.cpp
class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        vector<ListNode*> ans(k);
        ListNode* cur = head;
        int n = 0;

        while (cur){
            cur = cur->next;
            n++;
        }

        cur = head;
        int avg = n / k;
        int extra = n % k;

        for (int i = 0; i<k; i++){
            ans[i] = cur;
            int cnt = avg;

            if (extra-- > 0) cnt++;

            while (--cnt && cur) cur = cur->next;

            if (cur) {
                ListNode* prev = cur;
                cur = cur->next;
                prev->next = nullptr;
            }
        }
        return ans;
    }
};

File: 0725 Split Linked List in Parts/0725.py
class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        ans = [None] * k
        cur = head
        n = 0
        
        while cur:
            cur = cur.next
            n += 1
        
        avg = n // k
        extra = n % k
        cur = head

        for i in range(k):
            ans[i] = cur
            cnt = avg 

            if extra:
                cnt += 1
                extra -= 1

            while cur and cnt > 1:
                cur = cur.next
                cnt -= 1

            if cur:
                prev = cur
                cur = cur.next
                prev.next = None

        return ans
